ddl/ddl_api.go

```diff
@@ -5245,66 +5193,106 @@ func checkPlacementSpecConstraint(rules []*placement.Rule, rule *placement.Rule,
 			return rules, err
 		}
 
+		ruleCnt := int(replicas)
 		for labels, cnt := range constraints {
-			newRule := &placement.Rule{}
-			*newRule = *rule
+			newRule := rule.Clone()
 			if cnt <= 0 {
-				err = errors.Errorf("count should be non-positive, but got %d", cnt)
+				err = errors.Errorf("count should be positive, but got %d", cnt)
 				break
 			}
-			// TODO: handle or remove rule.Count in later commits
-			rule.Count -= cnt
+			if replicas != 0 {
+				ruleCnt -= cnt
+				if ruleCnt < 0 {
+					err = errors.Errorf("REPLICAS should be larger or equal to the number of total replicas, but got %d", replicas)
+					break
+				}
+			}
 			newRule.Count = cnt
-			err = checkPlacementLabelConstraints(newRule, strings.Split(strings.TrimSpace(labels), ","))
+
+			newRule.LabelConstraints, err = placement.CheckLabelConstraints(strings.Split(strings.TrimSpace(labels), ","))
 			if err != nil {
 				break
 			}
 			rules = append(rules, newRule)
 		}
+		rule.Count = ruleCnt
+
+		if rule.Count > 0 {
+			rules = append(rules, rule)
+		}
 	} else {
 		err = errors.Errorf("constraint should be a JSON array or object, but got '%s'", cnstr)
 	}
 	return rules, err
 }
 
-func checkPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
-	rules := make([]*placement.Rule, 0, len(specs))
+func buildPlacementSpecs(specs []*ast.PlacementSpec, tableID, partitionID int64) ([]*placement.RuleOp, error) {
+	rules := make([]*placement.RuleOp, 0, len(specs))
 
 	var err error
 	var sb strings.Builder
 	restoreFlags := format.RestoreStringSingleQuotes | format.RestoreKeyWordLowercase | format.RestoreNameBackQuotes
 	restoreCtx := format.NewRestoreCtx(restoreFlags, &sb)
 
 	for _, spec := range specs {
-		rule := &placement.Rule{
-			GroupID:  placementRuleDefaultGroupID,
-			Count:    int(spec.Replicas),
-			Override: true,
+		rule := &placement.RuleOp{
+			Rule: &placement.Rule{
+				GroupID:  placement.RuleDefaultGroupID,
+				Override: true,
```

Comment:
`Override` is not used.

Reply:
It is posted to PD for priorities. TiDB does not need to use this variable.